client/core/core.go

```diff
@@ -4563,16 +4565,40 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("match proof note unmarshal error: %w", err)
 	}
 
-	// Expire the epoch
-	dc.setEpoch(note.MarketID, note.Epoch+1)
+	// Expire previous epoch.
+	dc.tryUpdateEpoch(note.MarketID, note.Epoch+1)
 
 	book := dc.bookie(note.MarketID)
 	if book == nil {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	// Validate server match checksum for client orders (trades) in this epoch.
+	for _, trade := range dc.trackedTrades() {
+		// TODO - len chkeck, why ?
+		if trade.metaData.Status == order.OrderStatusEpoch && len(trade.csum) > 0 && !bytes.Equal(note.CSum, trade.csum) {
```

Comment:
Since epochs are continuous and there is time between epoch close and this match_proof msg (up to 20 sec for the preimage request timeout + network latency, although the spec incorrectly says 5 sec), there can be trades in `OrderStatusEpoch` from more than one epoch (different csums).

I'm afraid the `trackedTrade` doesn't know what epoch index it belongs to (sadly, server's `OrderResponse` doesn't include that), so finding the order with the corresponding preimage may be needed after all. However, that facilitates the lock-free solution to the `csum` data race issue.  I may be overlooking something...  @buck54321 I'm thinking cycling over preimages is gonna be needed after all, although I agree it doesn't quite warrant 2 helper functions.

As an aside, I'm really not sure if `dc.epoch` (and `setEpoch`) serves a purpose anymore, but let's not go down that rabbit how right now.  There are relics here and there we should note, but not change too readily.